### DISCUSSION TITLE
How to get window HWND?

### DISCUSSION THREAD

**Discussion by x000r** [emilk/egui#1887] [Q&A] [answered]
Hi, I'm trying to integrate `egui` with `vlc-rs`. VLC needs a window handle (HWND on windows) to use it for ouput. How do I get that? Are window containers in egui legit windows that can be referenced by HWND or are they accessible only from inside the framework itself (eframe)? I know I can use SDL, for example, to achieve the same, but I wanted to try egui out.

**Comment by emilk** [comment]
If you can get what you want from SDL, then perhaps try <https://crates.io/crates/egui_sdl2_gl>.

If you are using `eframe` however, there is nothing like that at the moment. `eframe` uses `winit` internally - does `winit` even expose it? A naive search does not come up with anything: https://docs.rs/winit/latest/winit/?search=hwnd

**Reply by x000r** [reply]
Thanks for the answer, understood, looks like I will have to play around sdl.

**Comment by coderedart** [comment, marked as answer]
> How do I get that?

what you are asking for is provided by the crate https://docs.rs/raw-window-handle/latest/raw_window_handle/index.html
ofcourse, all windowing crates implement the trait `HasRawWindowHandle` (glfw / sdl / winit). although, glfw also provides the handle directly too. 

> Are window containers in egui legit windows that can be referenced by HWND

Nope. egui (or eframe actually) only uses a single desktop window (atleast for now, might change if egui gets the MultiWindow feature https://github.com/emilk/egui/issues/552). 

>  are they accessible only from inside the framework itself (eframe)?

yep. the container windows are just egui areas pretending to be windows.

>  I know I can use SDL, for example, to achieve the same, but I wanted to try egui out.

from what i understand, VLC will draw on the whole window area that you give the HWND for. so, you cannot give the HWND of your main egui app, but instead create a child window with SDL integration and give VLC that child window's handle. ofcourse, it means your app and video are in separate windows.. 

Other alternative is to provide VLC your opengl context, which is very unsafe and idk if it would even work or whether the rust bindings provide the api for this. https://github.com/videolan/vlc/blob/master/doc/libvlc/sdl_opengl_player.cpp . but this would allow you to make vlc output to a texture and use that texture for a egui Image widget to display the output texture frames.

**Reply by x000r** [reply]
Appreciate your answer, things are more clear to me now, thanks.
